torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py

```diff
@@ -21,9 +22,20 @@ def __init__(
             **functional_optim_kwargs,
             _allow_empty_param_list=True,
         )
-        if not hasattr(self.functional_optimizer, "step_param"):
+        self._check_valid_functional_optim()
+
+    @classmethod
+    def from_functional_optim(cls, functional_optim):
```

Comment:
The difference between this function and the regular `__init__` is that this directly takes a functional optim class, but adding more arguments to the init and conditionally picking which one to use seemed a bit hacky.

From some online resources, it seems that one pythonic way around this is to use classmethods as here, hence this approach.

Reply:
A nit question that we don't have to address in this PR. Would it be a BC-breaking change if we just supported constructing `OptimizerHookState` directly from an optim class? What do we gain by the current `__init__` behavior beyond "injecting" our `_allow_empty_param_list` argument?